src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelBindingContext.cs

```diff
@@ -171,60 +57,24 @@ public string FieldName
         /// The <see cref="Model"/> will typically be set for a binding operation that works
         /// against a pre-existing model object to update certain properties.
         /// </remarks>
-        public object Model { get; set; }
+        public abstract object Model { get; set; }
 
         /// <summary>
         /// Gets or sets the metadata for the model associated with this context.
         /// </summary>
-        public ModelMetadata ModelMetadata
-        {
-            get { return _modelMetadata; }
-            set
-            {
-                if (value == null)
-                {
-                    throw new ArgumentNullException(nameof(value));
-                }
-
-                _modelMetadata = value;
-            }
-        }
+        public abstract ModelMetadata ModelMetadata { get; set; }
 
         /// <summary>
         /// Gets or sets the name of the model. This property is used as a key for looking up values in
         /// <see cref="IValueProvider"/> during model binding.
         /// </summary>
-        public string ModelName
-        {
-            get { return _modelName; }
-            set
-            {
-                if (value == null)
-                {
-                    throw new ArgumentNullException(nameof(value));
-                }
-
-                _modelName = value;
-            }
-        }
+        public abstract string ModelName { get; set; }
 
         /// <summary>
         /// Gets or sets the <see cref="ModelStateDictionary"/> used to capture <see cref="ModelState"/> values
```

Comment:
Yes, `<see cref="ModelState"/>` references the property it's describing and not the intended type.